### PR TITLE
Fix ignore side effects

### PR DESCRIPTION
**Pull request by AYM1607**
Fixes a bug where a composition is never synthesized if it was created ignoring side effects.

### CODE REVIEW

File: internal/controllers/synthesis/lifecycle.go [hunk 1]
```diff
@@ -363,12 +363,17 @@ func shouldSwapStates(synth *apiv1.Synthesizer, comp *apiv1.Composition) bool {
...
-	// - the bound input resources have changed
+	// - a side effect that has not observed a synthesis has occurred
+	//		The side effects observed by this controller are:
+	//			- changes to non-defferred inputs.
 	// AND
 	// - synthesis is not already pending
 	// - all bound input resources exist and are in lockstep (or composition is being deleted)
 	syn := comp.Status.CurrentSynthesis
-	return (comp.DeletionTimestamp != nil || (comp.InputsExist(synth) && !comp.InputsMismatched(synth))) && (syn == nil || syn.ObservedCompositionGeneration != comp.Generation || (!inputRevisionsEqual(synth, comp.Status.InputRevisions, syn.InputRevisions) && syn.Synthesized != nil))
+	return (syn == nil ||
```
Comment (jveski) [inline internal/controllers/synthesis/lifecycle.go:373]: So many parens it's starting to look like lisp 🥇 

Nice refactor, this was... not beautiful before.

File: internal/controllers/rollout/controller.go [hunk 2]
```diff
@@ -65,6 +65,20 @@ func (c *controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	// Do this before calculating selecting the next composition to avoid
+	// skewing the results.
+	for _, comp := range comps.Items {
+		if comp.Status.PendingResynthesis != nil && comp.ShouldIgnoreSideEffects() {
+			comp.Status.PendingResynthesis = nil
+			err := c.client.Status().Update(ctx, &comp)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("clearing PendingResythesis field for %s in namespace %s: %w", comp.Name, comp.Namespace, err)
+			}
+			return ctrl.Result{Requeue: true}, nil
```
Comment (jveski) [inline internal/controllers/rollout/controller.go:78]: Curious why we need to requeue here?

File: internal/controllers/synthesis/lifecycle.go (hunk 1 above)
Comment (AYM1607) [reply internal/controllers/synthesis/lifecycle.go:373]: It'll need more work, maybe in another PR...

File: internal/controllers/rollout/controller.go (hunk 2 above)
Comment (AYM1607) [reply internal/controllers/rollout/controller.go]: Good catch, we don't, we can rely on the write to kick off the next tick

File: examples/ignore-side-effects/main.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package main
```
Comment (jveski) [inline examples/ignore-side-effects/main.go:1]: nit: would prefer a toy shell script synth to avoid an extra image build during the smoke tests

**Review by jveski — APPROVED** [review]

File: examples/ignore-side-effects/main.go (hunk 3 above)
Comment (AYM1607) [reply examples/ignore-side-effects/main.go:1]: I'll track this in #208

**Pull request merged**